### PR TITLE
HBASE-24813 ReplicationSource should clear buffer usage on Replicatio…

### PR DESCRIPTION
**Pull request by wchevreuil**
This a backport for PR #2546 merged into master. The cherry-pick into branch-2 required some conflicts resolution.

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 13s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ branch-2 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   4m 10s |  branch-2 passed  |
| +1 :green_heart: |  checkstyle  |   1m 29s |  branch-2 passed  |
| +1 :green_heart: |  spotbugs  |   2m 44s |  branch-2 passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   4m 23s |  the patch passed  |
| -0 :warning: |  checkstyle  |   1m 27s |  hbase-server: The patch generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  hadoopcheck  |  15m 37s |  Patch does not cause any errors with Hadoop 3.1.2 3.2.1.  |
| +1 :green_heart: |  spotbugs  |   2m 47s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 16s |  The patch does not generate ASF License warnings.  |
|  |   |  43m 30s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2849/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/2849 |
| Optional Tests | dupname asflicense spotbugs hadoopcheck hbaseanti checkstyle |
| uname | Linux 3e177a1c62b0 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | branch-2 / cec9f18737 |
| checkstyle | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2849/1/artifact/yetus-general-check/output/diff-checkstyle-hbase-server.txt |
| Max. process+thread count | 95 (vs. ulimit of 12500) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2849/1/console |
| versions | git=2.17.1 maven=3.6.3 spotbugs=3.1.12 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 22s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  7s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ branch-2 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   4m 29s |  branch-2 passed  |
| +1 :green_heart: |  compile  |   1m 10s |  branch-2 passed  |
| +1 :green_heart: |  shadedjars  |   7m  8s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 45s |  branch-2 passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   4m 26s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 16s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 16s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   7m 24s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 43s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 138m 37s |  hbase-server in the patch passed.  |
|  |   | 169m 40s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2849/1/artifact/yetus-jdk11-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/2849 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux e385bb1f7404 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:06:04 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | branch-2 / cec9f18737 |
| Default Java | AdoptOpenJDK-11.0.6+10 |
|  Test Results | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2849/1/testReport/ |
| Max. process+thread count | 4373 (vs. ulimit of 12500) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2849/1/console |
| versions | git=2.17.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 29s |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  6s |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --whitespace-eol-ignore-list --whitespace-tabs-ignore-list --quick-hadoopcheck  |
||| _ Prechecks _ |
||| _ branch-2 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   4m 40s |  branch-2 passed  |
| +1 :green_heart: |  compile  |   1m 18s |  branch-2 passed  |
| +1 :green_heart: |  shadedjars  |   7m 13s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 41s |  branch-2 passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   4m 22s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 11s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 11s |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   7m 59s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 47s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 209m 25s |  hbase-server in the patch passed.  |
|  |   | 241m  3s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2849/1/artifact/yetus-jdk8-hadoop2-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/2849 |
| Optional Tests | javac javadoc unit shadedjars compile |
| uname | Linux 1536ea91ae89 4.15.0-126-generic #129-Ubuntu SMP Mon Nov 23 18:53:38 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | branch-2 / cec9f18737 |
| Default Java | AdoptOpenJDK-1.8.0_232-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2849/1/testReport/ |
| Max. process+thread count | 2887 (vs. ulimit of 12500) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2849/1/console |
| versions | git=2.17.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceShipper.java [hunk 1]
```diff
@@ -330,4 +333,56 @@ public boolean sleepForRetries(String msg, int sleepMultiplier) {
...
+        }
+      } catch (InterruptedException e) {
+        LOG.warn("{} Interrupted while waiting {} to stop on clearWALEntryBatch. "
+            + "Not cleaning buffer usage: {}", this.source.getPeerId(), this.getName(), e);
+        return;
+      }
+    }
+    LongAccumulator totalToDecrement = new LongAccumulator((a,b) -> a + b, 0);
+    entryReader.entryBatchQueue.forEach(w -> {
+      entryReader.entryBatchQueue.remove(w);
```
Comment (joshelser) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceShipper.java:374]: Is this iteration safe on a Queue? Usually, the safer thing to do is  to user an `iterator` on that `Collection` to remove the element as well. The `iterator.remove()` might also be more efficient.

**Review by joshelser — COMMENTED** [review]
One comment. Running tests locally to make sure we don't have a similar problem to last time :)

**Comment by joshelser** [comment]
> Running tests locally to make sure we don't have a similar problem to last time :)

Not observing any run-away logging for replication tests in hbase-server. 👍🏼

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceShipper.java (hunk 1 above)
Comment (wchevreuil) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceShipper.java:374]: My interpretation from _BlockingQueue_ javadoc is that it is: 

`BlockingQueue implementations are thread-safe. All queuing methods achieve their effects atomically using internal locks or other forms of concurrency control.`

**Review by ankitsinghal — APPROVED** [review]
Looks good to me.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceShipper.java (hunk 1 above)
Comment (joshelser) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceShipper.java:374]: Ah, yup. I'm also seeing that pop'ing off the front of the queue (at least in `ArrayBlockingQueue#remove`) is a fast operation. Thanks for thinking about this :)

**Review by joshelser — APPROVED** [review]

**Pull request merged**